### ISSUE TITLE
Problems with Gira USB interface and knxd

### ISSUE THREAD

**Issue by RalfM04** [knxd/knxd#371]
Hello, I've trouble to get knxd runninng.

First i tried to install the stable version on fresh Raspi 3b+ with actual stretch release.
After following the knxd install wiki from FHEM the service and daemon starting without problems.
After 1 or two switch operations the vbusmonitor write "read error" and the system starts again.
After reading and testing a lot about timing and delay and so on  I tried to start again with oldstable release under same conditions. Now the service didn't start and syslog shows following: "knxd[862]: initialisation of backend 'usb:' failed: Permission denied" I give up and search for some support.

**Comment by RalfM04** [comment]
Tried again,
using raspi strech lite latest.
Starting knxd with following command knxd -E 1.1.129:8 -D -T -R -S -b usb:
Result;
pi@fhem:~ $ systemctl status knxd
● knxd.service - KNX Daemon
   Loaded: loaded (/lib/systemd/system/knxd.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Thu 2018-11-22 19:48:25 CET; 4min 59s ago
  Process: 495 ExecStart=/usr/local/bin/knxd -p /run/knxd/knxd.pid $KNXD_OPTIONS (code=exited, status=1/FAILURE)

USB interface as listed below:
pi@fhem:~ $ findknxusb
Possible addresses for KNX USB devices:
device: 1:4:1:0:0 (Gira Giersiepen GmbH & Co. KG:KNX-USB Data Interface)

Version with applied patch
pi@fhem:~ $ knxd --v
knxd 0.14.25:cdbb66d-emipatch

Any idear ?

**Comment by tru7** [comment]
you are missing a least the -e parameter

**Comment by RalfM04** [comment]
Also tried this in the KNDX file
KNXD_OPTIONS="-e 1.1.128 -E 1.1.129:8 -D -T -R -S -b usb:"
Result;
pi@fhem:~ $ systemctl status knxd
● knxd.service - KNX Daemon
   Loaded: loaded (/lib/systemd/system/knxd.service; enabled; vendor preset: ena
   Active: failed (Result: exit-code) since Thu 2018-11-22 20:15:12 CET; 1min 13
  Process: 514 ExecStart=/usr/local/bin/knxd -p /run/knxd/knxd.pid $KNXD_OPTIONS

No idear whats going wrong

**Comment by smurfix** [comment]
May I ask why you are doing this?
Why is your knxd installed in /usr/local? Also, when using systemd nobody in their right mind would use a PID file.

Please add -t1022 -f9 (at the start of the options list) and reload. Also don't cut off the status lines ("systemctl status -l …"). 
One possible reason: your USB interface isn't listed in knxd's udev rules (or you didn't install them, or you didn't reload udev) and thus the USB interface's rights are wrong.

Also, what is "emipatch"? Where can I find it? Assuming that that patch makes sense, could you submit a pull request?

**Comment by RalfM04** [comment]
Answer I'm really not the Linux expert, so mostly I try to copy and paste from some explanations.
I tried to follow this one after I had no success with others before.
https://michlstechblog.info/blog/raspberry-pi-eibknx-ip-gateway-and-router-with-knxd/
This script asked to isntall the patchto expand the timing on the usb bus.
I read that sometimes this fixed the problem with the error after a few commands are read.
Now I trying to follow your explanations.
Wait a few minutes

**Comment by RalfM04** [comment]
Modified as follows:
KNXD_OTIONS="-t1022 -f9 -e 1.1.128 -E 1.1.129:8 -D -T -R -S -b usb:"
Result:
pi@fhem:~ $ systemctl status knxd
● knxd.service - KNX Daemon
   Loaded: loaded (/lib/systemd/system/knxd.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Thu 2018-11-22 21:13:11 CET; 2min 20s ago
  Process: 496 ExecStart=/usr/local/bin/knxd -p /run/knxd/knxd.pid $KNXD_OPTIONS (code=exited, status=2)

Nov 22 21:13:11 fhem systemd[1]: Starting KNX Daemon...
Nov 22 21:13:11 fhem systemd[1]: knxd.service: Control process exited, code=exited status=2
Nov 22 21:13:11 fhem systemd[1]: Failed to start KNX Daemon.
Nov 22 21:13:11 fhem systemd[1]: knxd.service: Unit entered failed state.
Nov 22 21:13:11 fhem systemd[1]: knxd.service: Failed with result 'exit-code'.

**Comment by smurfix** [comment]
Clarification: please remove the backslash, not the dollar sign.

**Comment by RalfM04** [comment]
Sorry I don't know what to do.
What means backslash escaped ?
Where should I remove this ?
Where is the knxd.service file located ?

To not waste your time I can setup the system with a clean install of raspian stretch lite and follow the install description without using this script and start knxd with systemd.
(That was what I tried at the beginning, till the system was terminatet after a while a starts again)

**Comment by tru7** [comment]
I guess your $KNXD_OPTIONS is just empty if you define KNXD_OTIONS=

**Comment by Race666** [comment]
The backslash is just in the install script. It quotes the $ sign while creating the service file with
```
cat >  /lib/systemd/system/knxd.service <<EOF
[Unit]
Description=KNX Daemon
After=network.target

[Service]
EnvironmentFile=/etc/default/knxd
# Wait for all interfaces, systemd-networkd-wait-online.service must be enabled
#ExecStartPre=/lib/systemd/systemd-networkd-wait-online --timeout=60
# Wait for a specific interface
#ExecStartPre=/lib/systemd/systemd-networkd-wait-online --timeout=60 --interface=eth0
ExecStart=/usr/local/bin/knxd -p /run/knxd/knxd.pid \$KNXD_OPTIONS
Type=forking
PIDFile=/run/knxd/knxd.pid
User=knxd
Group=knxd
#TimeoutStartSec=60

[Install]
WantedBy=multi-user.target network-online.target
EOF

```

**Comment by Race666** [comment]
Try to run knxd in foreground with your options
```
sudo -u knxd -s
knxd -t1022 -f9 -e 1.1.128 -E 1.1.129:8 -D -T -R -S -b usb:
```
and see whats happend in detail

**Comment by Race666** [comment]
> The values modified by that patch are configurable, so instead of patching code that works perfectly well for other devices one can simply add them as options …
@smurfix : Any hint to the documentation of these option?

**Comment by smurfix** [comment]
@race666 Ah. Thanks. I missed that. Ugh. Normal people use `<<'END'`  or whatever to quote the whole thing.

WRT documentation: `doc/inifile.rst` has them.

**Comment by Race666** [comment]
@smurfix Ok. Thanks
It is possible to set multiple parameters? Let us assume I want to set 
+  send_timeout = cfg->value("send-timeout",6000) / 1000.;
+  max_retries = cfg->value("send-retries",5);
like my patch. Then the correct syntax is
 knxd -t1022 -f9 -e 1.1.128 -E 1.1.129:8 -D -T -R -S **--arg=send-timeout=6000 --arg=send-retry=5**  -b usb:
Am I right?

**Comment by smurfix** [comment]
@Race666 Yes, that should work.

**Comment by smurfix** [comment]
> Try to run knxd in foreground
Note: before you do this, `systemctl stop knxd.socket`. Also add `-u -i` options, otherwise you won't be able to talk to knxd with `knxtool` (among others).

**Comment by Race666** [comment]
@smurfix "Normal people use <<'END' or whatever to quote the whole thing." -> **Normal poeple** ;-)
I agree with you quote all with 'EOL' is the better way in this case. In other cases when creating a file with cat I have both replaces the variable with its content and put a variable name in the new file. Then quoting the whole thing is not possible.

**Comment by RalfM04** [comment]
It seems running
Started knxtool vbusmonitor1 ip:localhost in another putty session
Receiving Telegramms by pressing any switch.

**Comment by smurfix** [comment]
"-u" doesn't work for you because you're not root and `/run` isn't writeable.

In any case, knxd itself works, so you need to find the bug in your systemd script / your configuration file. The documented way (build a knxd Debian packet from source and install that) works. I'd remove knxd from /usr/local and do what knxd's README says.

**Comment by RalfM04** [comment]
OK I will try that, I start again with a clean install following that way in the readme. Hopefully that way works, because then I'm back to the unpatched version. But of cause I feel also better If i could use the standart. 
I will give you a feedback after doing as descibed.

**Comment by RalfM04** [comment]
So, after long time of testing and installation it is done.
Raspian Stretch, acutal Master Branch and applied patch "increased default timeout and retry count for older usb interfaces #291".
All other atempts before leads to restart the knxd system. Reading was OK, writing leads to an error latest with the 5th command.
I'm really happy, but be carefull  - never change a runnig system ;-)

I really respect the work from smurfix and all others herem but for noobs like me it is hard to follow all the hints and suggestions.
Nevertheless, hopefully the current actual system does the same stable work as my raspberry 2 with eibd all the years before.
Thank you for your kindly support.
